File: Day 2/piloting-submarine.py
import numpy as np
from timer import timer
import logging

logging.basicConfig(level=logging.DEBUG)


def readfile():
    fname = 'input.txt'
    with open(fname, "r") as fd:
        directions = [x.split() for x in fd.read().splitlines()]
        directions = [[x[0], int(x[1])] for x in directions]
    return directions


@timer('function:partone', unit='ms')
def partone():
    dirs = [x if x[0] != "up" else [x[0], -x[1]] for x in readfile()]
    horizontal_position = np.sum([x[1] for x in dirs if x[0] == "forward"])
    depth = np.sum([x[1] for x in dirs if x[0] != "forward"])
    return horizontal_position * depth


@timer('function:parttwo', unit='ms')
def parttwo():
    directions = [x if x[0] != "up" else [x[0], -x[1]] for x in readfile()]
    horizontal, aim, depth = 0, 0, 0
    for direc in directions:
        if direc[0] != "forward":
            aim += direc[1]
        else:
            horizontal += direc[1]
            depth += aim * direc[1]
    return horizontal * depth


if __name__ == "__main__":
    print(partone())
    print(parttwo())

File: Day 2/piloting-submarine.go
package main
import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	log.Printf("%s took %s", name, elapsed)
}

func readFile() []string {
	fileName := "/Users/alex/main/advent-of-code/Day 2/input.txt"
	fileBytes, err := ioutil.ReadFile(fileName)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	navData := strings.Split(string(fileBytes), "\n")
	return navData
}

func partOne(navData []string) int {
	defer timeTrack(time.Now(), "partOne")
	hpoz, depth := 0, 0
	for i := range navData {
		lineData := strings.Split(string(navData[i]), " ")
		moveVal, _ := strconv.Atoi(lineData[1])
		if lineData[0] == "forward" {
			hpoz += moveVal
		} else if lineData[0] == "down" {
			depth += moveVal
		} else {
			depth -= moveVal
		}
	}
	return hpoz * depth
}

func partTwo(navData []string) int {
	defer timeTrack(time.Now(), "partTwo")
	hpoz, depth, aim := 0, 0, 0
	for i := range navData {
		lineData := strings.Split(string(navData[i]), " ")
		moveVal, _ := strconv.Atoi(lineData[1])
		if lineData[0] == "forward" {
			hpoz += moveVal
			depth += aim * moveVal
		} else if lineData[0] == "down" {
			aim += moveVal
		} else {
			aim -= moveVal
		}

	}
	return hpoz * depth
}

func main() {
	nData := readFile()
	fmt.Println(partOne(nData))
	fmt.Println(partTwo(nData))
}
